### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Matrix): characterize Jacobson radicals in matrix rings

### PR DESCRIPTION
**Pull request by Vtec234**
---

- [x] depends on: #17750

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 1]
```diff
@@ -5,20 +5,24 @@ Authors: Jujian Zhang, Wojciech Nawrocki
 -/
 import Mathlib.Data.Matrix.Basis
 import Mathlib.RingTheory.TwoSidedIdeal.Operations
+import Mathlib.RingTheory.JacobsonIdeal
```
Comment (Vtec234) [inline Mathlib/LinearAlgebra/Matrix/Ideal.lean:8]: This seems to be tagged with `large-import`. I initially split out the Jacobson results into a separate file, but decided it was too small to deserve a standalone existence.

Comment (YaelDillies) [reply Mathlib/LinearAlgebra/Matrix/Ideal.lean:8]: The impirt increase is *huge* though. Can you find a better way? maybe with a pereliminary split to an existing file?

Comment (Vtec234) [reply Mathlib/LinearAlgebra/Matrix/Ideal.lean:8]: What do you mean by 'preliminary split to an existing file'?

Comment (YaelDillies) [reply Mathlib/LinearAlgebra/Matrix/Ideal.lean:8]: Oh I hadn't noticed it xas a leaf file. Then >i'd say it's fine

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 2]
```diff
@@ -66,6 +70,47 @@ theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
...
+    ∀ x ∈ I.jacobson, ∀ (i j : n), stdBasisMatrix i j x ∈ (I.matricesOver n).jacobson := by
+  -- Proof generalized from example 8 in
+  -- https://ysharifi.wordpress.com/2022/08/16/the-jacobson-radical-basic-examples/
+  simp_rw [Ideal.mem_jacobson_iff]
+  intro x xIJ p q M
+  have ⟨z, zMx⟩ := xIJ (M q p)
+  let N : Matrix n n R := 1 - ∑ i, stdBasisMatrix i q (if i = q then 1 - z else (M i p)*x*z)
+  use N
+  intro i j
+  by_cases qj : q = j
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Matrix/Ideal.lean:92]:
```suggestion
  obtain rfl | qj := eq_or_ne q j
```

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 3]
```diff
@@ -66,6 +70,47 @@ theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
...
+  · by_cases ij : i = j
+    · rw [qj] at zMx
+      simp [qj, ij, N, zMx, stdBasisMatrix, mul_apply, sum_apply, ite_and, sub_mul]
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Matrix/Ideal.lean:95]:
```suggestion
    · simp [ij, N, zMx, stdBasisMatrix, mul_apply, sum_apply, ite_and, sub_mul]
```

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 4]
```diff
@@ -66,6 +70,47 @@ theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
...
+    · convert I.mul_mem_left (-M i p * x) zMx
+      simp [qj, ij, N, zMx, stdBasisMatrix, mul_apply, sum_apply, ite_and, sub_mul]
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Matrix/Ideal.lean:97]:
```suggestion
      simp [ij, N, zMx, stdBasisMatrix, mul_apply, sum_apply, ite_and, sub_mul]
```

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 5]
```diff
@@ -66,6 +70,47 @@ theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
...
+      simp [sub_add, mul_add, mul_sub, mul_assoc]
+  · simp [N, StdBasisMatrix.apply_of_col_ne _ _ qj, sum_apply, mul_apply]
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Matrix/Ideal.lean:99]:
```suggestion
  · simp [N, qj, sum_apply, mul_apply]
```

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 6]
```diff
@@ -66,6 +70,47 @@ theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
...
+
+/-- For any left ideal $I ≤ R$, we have $Mₙ(J(I)) ≤ J(Mₙ(I))$. -/
+theorem matricesOver_jacobson_le (I : Ideal R) :
+    I.jacobson.matricesOver n ≤ (I.matricesOver n).jacobson := by
+  intro M MI
+  rw [matrix_eq_sum_stdBasisMatrix M]
+  apply Ideal.sum_mem
+  intro i _
+  apply Ideal.sum_mem
+  intro j _
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Matrix/Ideal.lean:109]: This begs for a `membership` aesop-like tactic 😁

**Review by YaelDillies — COMMENTED** [review]
Thanks for the PR!

**Review by YaelDillies — APPROVED** [review]
maintainer merge

apart from that. Thanks!

File: Mathlib/LinearAlgebra/Matrix/Ideal.lean [hunk 7]
```diff
@@ -201,3 +245,34 @@ def orderIsoMatricesOver (i j : n) : TwoSidedIdeal R ≃o TwoSidedIdeal (Matrix
...
+  replace Mmem := mul_mem_right _ _ (stdBasisMatrix q p 1) Mmem
+  rw [mem_jacobson_iff] at Mmem
+  intro y
+  specialize Mmem (y • stdBasisMatrix p p 1)
+  have ⟨N, NxMI⟩ := Mmem
+  use N p p
+  simpa [mul_apply, stdBasisMatrix, ite_and] using NxMI p p
+
+/-- For any two-sided ideal $I ≤ R$, we have $J(Mₙ(I)) = Mₙ(J(I))$. -/
+theorem jacobson_matricesOver (I : TwoSidedIdeal R) :
```
[the remark this answers is not in the document]
Comment (Vtec234) [reply Mathlib/LinearAlgebra/Matrix/Ideal.lean:271]: Do you just mean
```lean
theorem jacobson_matricesOver_bot :
    ((⊥ : TwoSidedIdeal R).matricesOver n).jacobson =
    (⊥ : TwoSidedIdeal R).jacobson.matricesOver n :=
  jacobson_matricesOver ⊥
```
? I suppose I could, but it seems too trivial to extract as a standalone theorem?

[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/LinearAlgebra/Matrix/Ideal.lean:271]: Okay I might be hallucinating but is "matrices over not" just bot?

[the remark this answers is not in the document]
Comment (Vtec234) [reply Mathlib/LinearAlgebra/Matrix/Ideal.lean:271]: Oh yes I see, added something.

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Comment by riccardobrasca** [comment]
Can you please merge master and fix the error? Thanks!

bors d+

**Pull request closed without merging**